ksql-engine/src/main/java/io/confluent/ksql/metrics/Counter.java

```diff
@@ -0,0 +1,170 @@
+import org.apache.kafka.common.metrics.Metrics;
+import org.apache.kafka.common.metrics.Sensor;
+
+import java.text.SimpleDateFormat;
+import java.util.Collection;
+import java.util.Date;
+import java.util.Locale;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+class Counter<R> {
+
+  private final String topic;
+  private final Map<String, SensorMetric<R>> sensors;
+
+  Counter(String topic, final Map<String, SensorMetric<R>> sensors) {
+    this.topic = topic.toLowerCase();
+    this.sensors = sensors;
+  }
+
+  void increment(R record) {
+    sensors.values().forEach(v -> v.record(record));
+  }
+
+  public void close(Metrics metrics) {
+    sensors.values().forEach(v -> v.close(metrics));
+  }
+
+  boolean isTopic(String topic) {
+    return this.topic.equals(topic);
+  }
+
+  Collection<Stat> stats() {
+    return sensors.values().stream().map(sensor -> sensor.asStat()).collect(Collectors.toList());
+  }
+
+  public SensorMetric<R> firstSensor() {
+    if (sensors.isEmpty()) throw new RuntimeException("Sensors is empty");
+    return this.sensors.values().iterator().next();
+  }
+
+  static class Stat {
+    private final String name;
+    private double value;
+    private final long timestamp;
+
+
+    public Stat(String name, double value, long timestamp) {
+      this.name = name;
+      this.value = value;
+      this.timestamp = timestamp;
+    }
+    public String formatted() {
+      return  String.format("%s:%10.2f", name, value);
+    }
+
+    public String timestamp() {
+      return SimpleDateFormat.getDateTimeInstance(3, 1, Locale.getDefault()).format(new Date(timestamp));
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      if (this == o) return true;
+      if (o == null || getClass() != o.getClass()) return false;
+
+      Stat stat = (Stat) o;
+
+      if (Double.compare(stat.value, value) != 0) return false;
+      if (Double.compare(stat.timestamp, timestamp) != 0) return false;
+      return name != null ? name.equals(stat.name) : stat.name == null;
+    }
+
+    @Override
+    public int hashCode() {
+      int result;
+      long temp;
+      result = name != null ? name.hashCode() : 0;
+      temp = Double.doubleToLongBits(value);
+      result = 31 * result + (int) (temp ^ (temp >>> 32));
+      temp = Double.doubleToLongBits(timestamp);
+      result = 31 * result + (int) (temp ^ (temp >>> 32));
+      return result;
+    }
+
+
+    @Override
+    public String toString() {
+      return "Stat{" +
+              "name='" + name + '\'' +
+              ", value=" + value +
+              ", timestamp=" + timestamp +
+              '}';
+    }
+
+    public String name() {
+      return name;
+    }
+
+    public double getValue() {
+      return value;
+    }
+
+    public long getTimestamp() {
+      return timestamp;
+    }
+
+    public Stat aggregate(double value) {
+      this.value += value;
+      return this;
+    }
+  }
+
+  abstract static class SensorMetric<P> {
```

Comment:
Perhaps I have missed something, but why can't we just have a `List<Sensor>` inside the `Counter` class (or whatever we rename it to)? Each sensor already has a name. Each sensor already has a list of metrics. Each metric can have a particular type of Stat. 

So it seems to me that we can add utility functions to `Counter` to extract existing Kafka stats in a way that make them amenable to aggregation without introducing all these classes.

Reply:
Sensor is used to record() data and the KafkaMetric is used to read/metric.value() 's. You need both unfortunately. You also need a way of making them a bit more open to do aggs, and reregister properly.

Reply:
I guess this has been merged already, but it seems to me that the main public method in `TopicSensor` are `increment` and `stats`. 

If we didn't define `SensorMetric`, and had a list of `Sensor` instead, these could still be implemented as follows:

```
 void increment(R record) {
    sensors.forEach(v -> v.record(record));
 }

Collection<Stat> stats() {
    return sensors.stream().map(sensor -> extractValueFromSensor(sensor)).collect(Collectors.toList());
  }

private List<Stat> extractStatsFromSensor(Sensor sensor) {
   long now = time.now();
    sensor.metrics().stream().map(metric -> new Stat(metric.name(), metric.measurableValue(now), now);
}
```

It seems that `Stat` just helps associate metric names with values and time for aggregation. We probably could get rid of it too, since all of the data is in the underlying metric. We would just move the `extractStatsFromSensor` logic to the point of aggregation. 

This would help simplify a little bit, but doesn't really change any of the logic.

Reply:
oh man - why didnt I see the metrics() call sensor - i thought it was daft and instead it was me being blind.

Reply:
:) Yea, it is easy to miss. I spent a lot of time digging through the Kafka metrics code because it seemed to me that we shouldn't have to write these types of classes in general.